app/services/hud_reports/s3_artifact_service.rb

```diff
@@ -0,0 +1,293 @@
+###
+# Copyright 2016 - 2025 Green River Data Analysis, LLC
+#
+# License detail: https://github.com/greenriver/hmis-warehouse/blob/production/LICENSE.md
+###
+
+# frozen_string_literal: true
+
+# Service for managing HUD report artifacts in S3 via ActiveStorage
+# Handles storing report data as CSV files and JSON summaries
+module HudReports
+  class S3ArtifactService
+    include ActionView::Helpers::DateHelper
+
+    def initialize(report_instance)
+      @report = report_instance
+    end
+
+    # Store all report artifacts to S3
+    def store_artifacts!
+      # Prefer sharded universe members per question to reduce read size
+      store_universe_members_csv_shards!
+      store_report_clients_csv!
+      store_cell_details_csv!
+      store_report_summary_json!
+    end
+
+    # Store universe members as per-question shards
+    def store_universe_members_csv_shards!
+      # Avoid duplicating work if shards already exist
+      return if @report.universe_members_csv_shards.attached? && @report.universe_members_csv_shards.attachments.any?
+
+      questions = @report.report_cells.distinct.pluck(:question)
+      questions.each do |question|
+        csv_data = generate_universe_members_csv_for_question(question)
+        filename = "universe_members_#{@report.id}_#{question}_#{Time.current.to_fs(:number)}.csv"
```

Comment:
These filenames should probably include the report class or table name.  Once we start adding non-HUD reports, there might be collisions with the ids.  Maybe:

```suggestion
        filename = "universe_members_#{@report.class.table_name}_#{@report.id}_#{question}_#{Time.current.to_fs(:number)}.csv"
```